lib/pacemaker/pcmk_graph_logging.c

```diff
@@ -169,16 +182,20 @@ static void
 log_synapse(unsigned int log_level, pcmk__graph_t *graph,
             pcmk__graph_synapse_t *synapse)
 {
-    char *pending = NULL;
+    GString *pending = NULL;
 
     if (!pcmk_is_set(synapse->flags, pcmk__synapse_executed)) {
```

Comment:
Is this even right? Based on `synapse_state_str()`, a synapse with the executed flag set can be in-flight.

Also, as the coverity error pointed out, we'll currently dereference a null pointer if the executed flag is set (so we skip this block) and then we enter the for-loop.

It seems like we'd want to either:
* enclose everything inside the `if (!pcmk_is_set(synapse->flags, pcmk__synapse_executed))` block (for whatever reason, not sure why it exists) so that `pending` is non-NULL when the for-loop runs;
* assign `"none"` to `pending` if the executed bit is set... which seems wrong, since it could be in-flight; or
* take the `pending = synapse_pending_inputs(graph, synapse)` line out of the if-block so that we run it no matter what.

I lean toward option 3 but there might be a reason things are the way they are.

Reply:
We did lose a null check on `pending` in 5b33ce3; previously we logged `"none"` if `pending == NULL`. That's a trivial enough fix.

However, even before that commit, we ran the equivalent of `synapse_pending_inputs()` only if `synapse->executed` was false. But `synapse_pending_inputs()` considers in-flight actions (based on the `// In-flight or pending` comment), and it sounds like a synapse with `executed` set to true could have in-flight actions... so I'm not sure why we'd want to guard the `synapse_pending_inputs()` call inside `if !pcmk_is_set(synapse->flags, pcmk__synapse_executed))`.

Reply:
> But it's odd that coverity didn't complain about the original:

free() accepts (and ignores) NULL, but most glib functions don't (the GList functions are an exception since NULL is a valid empty GList)

Reply:
Coverity complains about dereferencing a null pointer here:
```
        log_synapse_action(log_level, synapse,
                           (pcmk__graph_action_t *) lpc->data,
                           (const char *) pending->str);
```
not in the free

Reply:
> It seems like we'd want to either:

The current code is just a style preference. Some people prefer to have as little indentation as possible. But even better for that purpose would be to do "if (pcmk_is_set(...)) { return; }" at the start. Enclosing the entire thing in "if (!pcmk_is_set(...))" is also a reasonable approach.

Reply:
> However, even before that commit, we ran the equivalent of synapse_pending_inputs() only if synapse->executed was false. But synapse_pending_inputs() considers in-flight actions (based on the // In-flight or pending comment), and it sounds like a synapse with executed set to true could have in-flight actions

I was failing to differentiate between the synapse being in-flight versus its pending actions being in-flight. The if-guard makes more sense now.

Reply:
> Some people prefer to have as little indentation as possible.

I prefer to return/continue/break early to have less indentation myself

Reply:
> We did lose a null check on pending in https://github.com/ClusterLabs/pacemaker/commit/5b33ce3995ce08d68519c5c329991ef9010fba9c; previously we logged "none" if pending == NULL. That's a trivial enough fix.

synapse_pending_inputs() returns "none" for NULL now

Reply:
> I guess because the dereference is now explicit in this function instead of in whatever qb function do_crm_log calls

Yeah you're right, either log_synapse_action() should use pcmk__s() on pending_inputs, or pending should be strdup("none") if synapse_pending_inputs() isn't called

Reply:
> synapse_pending_inputs() returns `"none"` for NULL now

Which is great, if `synapse_pending_inputs()` gets called :)

The latest update gets rid of the `strdup("none")`/`g_string_new("none")` altogether, sets a `const char *` to "none" by default in `log_synapse()`, and updates it to point to the pending list if `synapse_pending_inputs()` is called and returns non-`NULL`.